src/fundus/scraping/pipeline.py

```diff
@@ -1,47 +1,163 @@
+
+        """
+        ...
+
+
 class Pipeline:
     def __init__(self, *scrapers: Scraper):
+        """Basic pipeline to utilize scrapers.
+
+        Because scrapers are implemented asynchronous this pipeline handles
+        the necessary event loops and program logic to download articles
+        in batches asynchronously.
+
+        Args:
+            *scrapers (Scraper): The scrapers which should be used.
+        """
         self.scrapers: Tuple[Scraper, ...] = scrapers
 
     def run(
         self,
         error_handling: Literal["suppress", "catch", "raise"],
         max_articles: Optional[int] = None,
         extraction_filter: Optional[ExtractionFilter] = None,
-        batch_size: int = 10,
+        delay: Optional[Delay] = lambda: 0.1,
+        url_filter: Optional[URLFilter] = None,
+        only_unique: bool = True,
     ) -> Iterator[Article]:
-        scrape_map = map(
-            lambda x: x.scrape(
-                error_handling=error_handling, batch_size=batch_size, extraction_filter=extraction_filter
-            ),
-            self.scrapers,
-        )
-        robin = more_itertools.interleave_longest(*tuple(scrape_map))
-
-        if max_articles:
-            while max_articles:
-                try:
-                    yield next(robin)
-                except StopIteration:
-                    pass
-                max_articles -= 1
+        """Yields articles from initialized scrapers.
+
+        The articles will be requested concurrently with batch size := len(self.scrapers).
+        You can specify a Delay to be used between the batches with <delay>
+
+        Works like a light-wight version of Crawler.crawl(). Ment to be used when dealing with
+        custom scrapers outside the context of predefined publisher collections.
+        Refer to the docstring of Crawler.crawl() for more detailed information about the Args.
+
+        Args:
+            max_articles (Optional[int]): Maximal number of articles to be yielded. Defaults to None.
+            error_handling (Literal["suppress", "catch", "raise"]): Set error handling
+                for extraction. Defaults to "suppress".
+            extraction_filter (Optional[ExtractionFilter]): Set extraction filter. Defaults to None.
+            delay (Optional[Delay]): Set waiting time between article batches. Defaults to None.
+            url_filter (Optional[URLFilter]): Set URLFilter. Defaults to None
+            only_unique: (bool): If true return only unique responses. Defaults to True.
+
+        Returns:
+            Iterator[Article]: An iterator yielding objects of type Article.
+        """
+
+        response_cache: Set[str] = set()
+
+        for scraper in self.scrapers:
+            for source in scraper.sources:
+                if url_filter:
+                    source.add_url_filter(url_filter=url_filter)
+                if only_unique:
+                    source.add_url_filter(url_filter=lambda url: url in response_cache)
```

Comment:
Instead of mutating the `url_filter` from the initialization, we could also accept an additional `url_filter` in the `async_fetch` method. Then the union of both filters would be used.

Reply:
It will be moved to an issue.